truss/remote/baseten/remote.py

```diff
@@ -181,16 +200,135 @@ def push(  # type: ignore
             truss_handle=truss_handle,
             api=self._api,
         )
 
+    def push_chain_atomic(
+        self,
+        chain_name: str,
+        chainlet_push_payloads: List[definitions.ChainletPushPayload],
+        publish: bool = False,
+        environment: Optional[str] = None,
+    ):
+        if environment:
+            # If we are promoting a model to an environment after deploy, it must be published.
+            # Draft models cannot be promoted.
+            publish = True
+
+        chain_id = get_chain_id_by_name(self._api, chain_name)
+        chainlet_data: List[custom_types.ChainletDataAtomic] = []
+        entrypoint_truss_handle: Optional[TrussHandle] = None
+
+        for push_payload in chainlet_push_payloads:
+            if push_payload.is_entrypoint:
+                entrypoint_truss_handle = push_payload.truss_handle
+
+            options = self._prepare_push(
+                truss_handle=push_payload.truss_handle,
+                model_name=push_payload.model_name,
+                # Models must be trusted to use the API KEY secret.
+                trusted=True,
+                publish=publish,
+                origin=custom_types.ModelOrigin.CHAINS,
+            )
+
+            oracle_data = custom_types.OracleData(
+                name=push_payload.name,
+                model_name=options.model_name,
+                s3_key=options.s3_key,
+                config=options.encoded_config_str,
+                client_version=f"truss=={truss.version()}",
+                is_draft=options.is_draft,
+                model_id=options.model_id,
+                is_trusted=options.is_trusted,
+                preserve_previous_prod_deployment=options.preserve_previous_prod_deployment,
+                deployment_name=options.deployment_name,
+                origin=options.origin,
+                environment=options.environment,
+            )
+
+            chainlet_data.append(
+                custom_types.ChainletDataAtomic(
+                    name=push_payload.name,
+                    is_entrypoint=push_payload.is_entrypoint,
+                    oracle=oracle_data,
+                )
+            )
+
+        assert entrypoint_truss_handle is not None
+
+        chain_deployment_handle = create_chain_atomic(
+            api=self._api,
+            chain_id=chain_id,
+            chain_name=chain_name,
+            chainlets=chainlet_data,
+            is_draft=not publish,
+            environment=environment,
+        )
+
+        model_id = chain_deployment_handle.entrypoint_model_id
+        model_version_id = chain_deployment_handle.entrypoint_model_version_id
+
+        entrypoint_service = BasetenService(
+            model_id=model_id,
+            model_version_id=model_version_id,
+            is_draft=options.is_draft,
+            api_key=self._auth_service.authenticate().value,
+            service_url=f"{self._remote_url}/model_versions/{model_version_id}",
```

Comment:
I don't know that that's necessary because `chains_remote.push` ultimately returns a `BasetenChainService` which exposes the correct [Chain invocation URL](https://github.com/basetenlabs/truss/blob/694fe00b8c24dfa9716d4e6b4c0b0944c45da94e/truss-chains/truss_chains/remote.py#L245-L251).

Reply:
Hmm, maybe it's a bit unclean to return a (model-centric) `BastenService` here. We could either try to directly create `BasetenChainService` here or only return `chain_deployment_handle`. What's the point of the Model service?

Reply:
Hm yeah this is a bit tricky. `BasetenChainService` inherits from `ChainService` which requires an entrypoint service (instance of `BastenService`). Also, `BasetenChainService` is defined in `truss-chains` so I didn't want to create a reverse dependency. I think changing this would require a larger refactor.